### ISSUE TITLE
Listening for same event recurringly does not resolve latest non-canceled task

### ISSUE THREAD

**Issue by nilsmehlhorn** [microsoft/durabletask-dotnet#147]
The docs of [TaskOrchestrationContext.WaitForExternalEvent](https://learn.microsoft.com/en-us/dotnet/api/Microsoft.DurableTask.TaskOrchestrationContext.WaitForExternalEvent?view=durabletask-dotnet-1.x&viewFallbackFrom=netstandard-2.0) state:

> Orchestrators can wait for the same event name multiple times, so waiting for multiple events with the same name is allowed. Each external event received by an orchestrator will complete just one task returned by this method.

However, when cancelling an earlier listener task before any event with that name has arrived, a later listener task won't be resolved by the next arriving event.

I've created a minimal reproduction in [this repository](https://github.com/nilsmehlhorn/DurableFunctionMultiEventRepro), specifically inside [MyOrchestration.cs](https://github.com/nilsmehlhorn/DurableFunctionMultiEventRepro/blob/main/MyOrchestration.cs) where a listener for event name "RecurringEvent" is registered three times while only one listener is active (meaning not-cancelled) at a time and only the last listener is meant to resolve. 

After starting the orchestrator via http://localhost:7071/api/MyOrchestration_HttpStart, I'd expect the following steps to result in its completion:

1. Raise event "OneTimeEvent1"
2. Raise event "OneTimeEvent2"
3. Raise event "RecurringEvent"

Instead, the orchestrator only completes after sending "RecurringEvent" one more time:

1. Raise event "OneTimeEvent1"
2. Raise event "OneTimeEvent2"
3. Raise event "RecurringEvent"
4. Raise event "RecurringEvent"

At first I thought the runtime might serve earlier listeners even though they're canceled, however, this does not seem to be exactly the case since I'm registering three listeners while its sufficient to send the corresponding event only two times.

## System Info

- MacBook Pro M1 Ventura 13.2.1 / Windows 10 on Desktop PC
- Azurite for Storage Emulation

## Workaround

I've refactored code that relied on this pattern to hold a reference to a single listener instead of cancelling earlier listeners and starting new ones.

**Comment by jviau** [comment]
I was able to get a minimal repro of this. The issue is that we process pending events in a FIFO manner but cancelling _does not dequeue_ that event. When the actual event comes in, we try to complete the already cancelled event _then dequeue it_, so the later events will actually get to the latest non-cancelled task.

Removing those pending events on cancellation is easy enough, but I need to think through the repercussions of it a bit more. It is technically a behavior breaking change to address this now. But I am also wondering if there are scenarios where this is desired behavior?

Your workaround for now is for the best - re-using that task will avoid any framework behavior confusion.

**Comment by jviau** [comment]
@cgillum do you want to weigh in on what you think is the desired behavior here?

I have a minimal repro integration test here: https://github.com/jviau/durabletask-dotnet/blob/event-issue/test/Grpc.IntegrationTests/CancelledEventTest.cs

**Comment by cgillum** [comment]
I think it makes sense to remove canceled tasks from the list so that future events only trigger non-canceled tasks. That feels a little more intuitive to me.

**Comment by jviau** [comment]
Yeah, that makes sense to me also. However, something about the event pattern is still not ideal when it comes to cancellations and re-subscribing to the same event name. It introduces potential race conditions. Consider the orchestration in the original issue, if you were to start that orchestration and then immediately send those 3 events in quick succession (or start it with those 3 events), then it is possible the `RecurringEvent` will complete one of the first wait's on it, _before cancellation occurs_. And so you would still need to send multiple `RecurringEvent`'s to this orchestration to unblock it.

Essentially, for @nilsmehlhorn orchestration to work, you must have insight into the current progress of the orchestration and only send the "RecurringEvent" when it has already progressed past cancelling the existing wait handles. The current behavior is more deterministic - regardless of orchestration progress, you will always have to send the same number of repeat events to unblock the orchestration.

I wonder if there is a way to make this more deterministic. Can we wait on resolving the event until it is actually `awaited`? This way the early event would never be consumed by the yet-to-be-cancelled wait handles, because they are always cancelled before awaited. After putting this to writing I don't know how valuable that would actually be. I think the primary takeaway for me is that event behavior is highly dependent on order they come in relative to orchestration progress. Is there anything we can do about that?

**Comment by cgillum** [comment]
> regardless of orchestration progress, you will always have to send the same number of repeat events to unblock the orchestration.

I'm not sure that's necessarily better. Whether an orchestration completes after 3 events depends on whether that's the intended behavior of the orchestration (it clearly wasn't the intended behavior in this scenario). I'm not super comfortable describing one approach as more "deterministic" than the other.

One scenario I potentially worry about is one where you receive external events of a particular name in a loop. Something like this:

```csharp
CancellationTokenSource cts = new();
// ...
while (true)
{
    try
    {
        await context.WaitForExternalEventAsync("MyEvent", cts.Token);
        break;
    }
    catch (OperationCanceledException)
    {
        // log something, report a custom status, call an activity, etc.
    }
}
```

If the intent is to take some periodic action while waiting for an event to arrive, you might write code like the above. With the current behavior, you more-or-less have a task leak because each call to `WaitForExternalEventAsync` creates a new task and leaves all the existing tasks intact. That feels bad to me. I'd feel better if the code above could break out of the loop after receiving just one event.

I didn't actually look closely at the repro example to see whether the problem would occur in this case, so I apologize if I'm not understanding the issue fully. My hope is that the above works in an intuitive way.

**Comment by jviau** [comment]
I am not saying the current behavior is _better_ outright, just pointing out that determinism is a benefit in its own right. But I don't think we should rush into making a change to dequeue the tasks on cancellation without further consideration. I think that will introduce its own set of problems (on top of being a behavioral breaking change). What I would like is for us to think deeper about the scenario and understand what canceling and requeuing the same event wait handle repeatedly really implies.

I also am starting to wonder what the point of cancellation tokens here is. Trying to think of a scenario where it is beneficial over an alternative. There are so many unknowns with cancellation tokens in durable right now. For one, they won't wake up your orchestration when fired - so using external cancellation tokens won't wake up an orchestration without some other history event occurring. The other is that with the way durable drops continuations, they are potential leaks (any `IDisposables` instantiated in an orchestration may not have `Dispose` called on it - even with `using` statements. Need to verify this). In the end you only can really using cancellation tokens from sources you own and have been created during your `RunAsync`. In which case, why not use timers and `Task.WhenAny`? This way you will actually be woken up and don't need to worry about leaks.

Take your sample for example, it would be preferrable to use a timer + `Task.WhenAny` to perform the same. No need to complete the original wait handle, just wake up on an interval to log it has not completed.

``` CSharp
Task externalEvent = context.WaitForExternalEventAsync<object>("MyEvent");
while (true)
{
    Task completed = await Task.WhenAny(externalEvent, context.Delay(context.CurrentDateTimeUtc + someDelay));
    if (completed == externalEvent)
    {
        break;
    }
    
    // log something, report a custom status, call an activity, etc.
}
```

**Comment by nilsmehlhorn** [comment]
If I may weigh in with a perspective from userland:

> I also am starting to wonder what the point of cancellation tokens here is.

If the behavior is somewhat intended, I'd also say that waiting for external events shouldn't accept cancellation tokens. That's also how the API works when not using the isolated mode (we discovered the bug during migration) - at least when you're not _also_ using timers.

> [...] external cancellation tokens won't wake up an orchestration without some other history event occurring.

For context, we're using external events to fire our cancellation tokens in an effort to model the graceful cancellation of our business process (i.e. order fulfillment). The API around this basically looks like follows (after workaround was implemented):

```csharp
// registers event listener under the hood
var cancellation = new OrchestratorCancellation(context);
// races cancellable task against external event via Task.WhenAny
await cancellation.Wrap(cancellationToken => RunMyCancellableTaskAsync(cancellationToken));
```

I'd agree that `Task.WhenAny` and/or timers are the right move in orchestrators, however, the API in conjunction with loops and if statements is a bit clumsy. Due to this, we're actually considering to refactor certain orchestrators into entities which may then leverage state machine patterns/frameworks.

**Comment by jviau** [comment]
> That's also how the API works when not using the isolated mode 

@nilsmehlhorn Can you clarify this? Do you mean the in-proc API does not accept cancellation tokens, or that it does dequeue the task when cancellation is triggered?

 I am interested in seeing how you want to use cancellation tokens, and what code patterns it allows you to do that is not doable with timers (or is preferrable over the timer approach).

As for graceful/external orchestration cancellation, that is something we want to solve, but that work has not started yet.

> we're actually considering to refactor certain orchestrators into entities

Durable entities are not yet supported in isolated. We are working on it though.

**Comment by nilsmehlhorn** [comment]
@jviau I meant that the in-proc API does not accept cancellation tokens - at least if you're note also passing a timer and even then I seem to recall that the behavior was similar regarding the missing dequeue.

> I am interested in seeing how you want to use cancellation tokens, and what code patterns it allows you to do that is not doable with timers (or is preferrable over the timer approach).

Right now we're only using custom cancellation tokens based on external events.
